src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,217 @@
+        redis_password: Option<String>,
+        is_tls: bool,
+    ) -> Res<PrivacyBudget> {
+        //if Redis server needs secure connection
+        let uri_scheme = if is_tls { "rediss" } else { "redis" };
+
+        let redis_conn_url = format!(
+            "{}://:{}@{}",
+            uri_scheme,
+            redis_password.unwrap_or_default(),
+            redis_host_name
+        );
+
+        let redis_connection = redis::Client::open(redis_conn_url)?.get_connection()?;
+
+        Ok(PrivacyBudget {
+            redis_connection,
+            total_privacy_budget: 100,
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys that are received.
+    /// For those that end up exceeding the budget, the budget is then decremented again,
+    /// but the corresponding match key is discarded and not passed to the output.
+    pub fn get_eligible_packets_with_budget(
+        &mut self,
+        packets: &[BlindedMatchKey],
+        budget: u32,
+    ) -> Vec<BlindedMatchKey> {
+        let mut eligible_packets: Vec<BlindedMatchKey> = Vec::new();
+        for packet in packets.iter() {
+            let _eligible = match self.increase_budget_consumed(packet, budget) {
+                Ok(true) => {
+                    eligible_packets.push(*packet);
+                    Ok(())
+                }
+                Ok(false) => self.decrease_budget_consumed(&*packet, budget),
+                Err(_) => Ok(()),
+            };
+        }
+        eligible_packets
+    }
+
+    pub fn get_total_privacy_budget(&self) -> u32 {
+        self.total_privacy_budget
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
+        let key = packet.to_string();
+        let (budget_consumed,): (u32,) =
+            transaction(&mut self.redis_connection, &[key.clone()], |con, pipe| {
+                let val: u32 = con.get(&key).unwrap_or(0);
+                // increment
+                pipe.set(&key, val + budget).ignore().get(&key).query(con)
+            })?;
+        if budget_consumed > self.get_total_privacy_budget() {
+            Ok(false)
+        } else {
+            Ok(true)
+        }
+    }
+
+    /// Budget is decreased for each ineligible key in Redis
+    fn decrease_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<()> {
+        self.redis_connection.decr(packet.to_string(), budget)?;
+        Ok(())
+    }
+}
+
+impl From<RedisError> for Error {
+    fn from(_e: RedisError) -> Self {
+        Error::RedisError
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::thread_rng;
+    use std::sync::Mutex;
+
+    lazy_static! {
+        static ref PRIVACY_BUDGET: Mutex<PrivacyBudget> =
+            Mutex::new(PrivacyBudget::new("127.0.0.1", None, false).unwrap());
+    }
```

Comment:
This is basically OK, but you need to ensure that you cleanup properly or this database will accumulate a lot of old values.  Not a problem in CI, but local runs that fail will leave stuff behind.  Implementing Drop on the test keys you use might help with that.

Reply:
Since privacy budget is a mutex, Drop was giving an issue due to over-locking. Have dropped the lock before each assert so the destructor is able to clear the redis keys